lib/vnet/vnet.go

```diff
@@ -0,0 +1,441 @@
+func (c *Config) CheckAndSetDefaults() error {
+	if c.TUNDevice == nil {
+		return trace.BadParameter("TUNdevice is required")
+	}
+	if c.IPv6Prefix.Len() != 16 || c.IPv6Prefix.AsSlice()[0] != 0xfd {
+		return trace.BadParameter("IPv6Prefix must be an IPv6 ULA address")
+	}
+	return nil
+}
+
+// IPv6Prefix returns a Unique Local IPv6 Unicast Address which will be used as a 64-bit prefix for all v6 IP
+// addresses in the VNet.
+func IPv6Prefix() (tcpip.Address, error) {
+	// |   8 bits   |  40 bits   |  16 bits  |          64 bits           |
+	// +------------+------------+-----------+----------------------------+
+	// | ULA Prefix | Global ID  | Subnet ID |        Interface ID        |
+	// +------------+------------+-----------+----------------------------+
+	// ULA Prefix is always 0xfd
+	// Global ID is random bytes for the specific VNet instance
+	// Subnet ID is always 0
+	// Interface ID will be the IPv4 address prefixed with zeros.
+	var bytes [16]byte
+	bytes[0] = 0xfd
+	if _, err := rand.Read(bytes[1:6]); err != nil {
+		return tcpip.Address{}, trace.Wrap(err)
+	}
+	return tcpip.AddrFrom16(bytes), nil
+}
+
+// TUNDevice abstracts a virtual network TUN device.
+type TUNDevice interface {
+	// Write one or more packets to the device (without any additional headers).
+	// On a successful write it returns the number of packets written. A nonzero
+	// offset can be used to instruct the Device on where to begin writing from
+	// each packet contained within the bufs slice.
+	Write(bufs [][]byte, offset int) (int, error)
+
+	// Read one or more packets from the Device (without any additional headers).
+	// On a successful read it returns the number of packets read, and sets
+	// packet lengths within the sizes slice. len(sizes) must be >= len(bufs).
+	// A nonzero offset can be used to instruct the Device on where to begin
+	// reading into each element of the bufs slice.
+	Read(bufs [][]byte, sizes []int, offset int) (n int, err error)
+
+	// BatchSize returns the preferred/max number of packets that can be read or
+	// written in a single read/write call. BatchSize must not change over the
+	// lifetime of a Device.
+	BatchSize() int
+
+	// Close stops the Device and closes the Event channel.
+	Close() error
+}
+
+// Manager holds configuration and state for the VNet.
+type Manager struct {
+	tun           TUNDevice
+	stack         *stack.Stack
+	linkEndpoint  *channel.Endpoint
+	ipv6Prefix    tcpip.Address
+	rootCtx       context.Context
+	rootCtxCancel context.CancelFunc
+	state         state
+	slog          *slog.Logger
+}
+
+type state struct {
+	mu               sync.RWMutex
+	tcpHandlers      map[tcpip.Address]tcpHandler
+	nextFreeIPSuffix uint32
+}
+
+func newState() state {
+	return state{
+		tcpHandlers:      make(map[tcpip.Address]tcpHandler),
+		nextFreeIPSuffix: 100<<24 + 64<<16 + 0<<8 + 2<<0,
+	}
+}
+
+// tcpConnector is a type of function that can be called to consume a TCP connection.
```

Comment:
I took a stab at what I had in mind here: https://github.com/gravitational/teleport/compare/nklaassen/vnet0...isaiah/vnet0-suggestion?expand=1

The basic idea was to try and make the flow of a `tcp.ForwardRequest` through the system more readily apparent via the types. It also incidentally eliminated the need for anything related to our discussion [here](https://github.com/gravitational/teleport/pull/40889#discussion_r1581733072). 

Your call whether to take it or leave it.

Reply:
I do appreciate the detailed review, and I'm not ignoring this comment, I'm just working out the details of this API at the top of the PR stack where I start to see the real implementation, coming soon ™️

Reply:
I'm iterating on this in https://github.com/gravitational/teleport/pull/41031

There's a reason for doing the thing where the handler has to call a function to get the `net.Conn`:
```
// TCPHandler defines the behavior for handling TCP connections from VNet.
//
// Implementations should attempt to dial the target application and return any errors before calling
// [connector] to complete the TCP handshake and get the TCP conn. This is so that clients will see that the
// TCP connection was refused, instead of seeing a successful TCP dial that is immediately closed.
type TCPHandler interface {
	HandleTCP(ctx context.Context, connector func() (net.Conn, error)) error
}
```